File: apps/ios/CoreHub/Phone/Locate.swift
// An agent asks where this phone is (contract decision §103): the request arrives on `/rt/devices`
// (`request.created`, sent to this device only) or is caught up with `listRequests?status=pending`
// when the app comes back; the person is asked once — «Allow every time», «Only this time»,
// «Don't allow» — and iOS's own location permission follows. «Every time» and «never» are kept on
// the phone and changed under This device; the hub only ever gets one answer per request.
import CoreHubClient
import CoreLocation
import Foundation
import Observation
import SwiftUI

enum LocationChoice: String, CaseIterable, Identifiable {
    case ask, always, never
    var id: String { rawValue }
}

/// The rules of answering, apart from iOS so they are unit-tested.
enum Locating {
    enum Next: Equatable { case ask, answer, deny, ignore }

    static func next(_ request: DeviceRequest, choice: LocationChoice, now: Date = Date()) -> Next {
        guard request.capability == .location, request.status == .pending, request.expiresAt > now else { return .ignore }
        switch choice {
        case .never: return .deny
        case .always: return .answer
        case .ask: return .ask
        }
    }

    /// The one shape a location has (§14).
    static func result(latitude: Double, longitude: Double, accuracy: Double, at date: Date) -> [String: JSONValue] {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        return [
            "latitude": .double(latitude),
            "longitude": .double(longitude),
            "accuracy_m": .double(max(0, accuracy)),
            "captured_at": .string(formatter.string(from: date)),
        ]
    }

    static func fulfilled(_ result: [String: JSONValue]) -> DeviceRequestResponse { DeviceRequestResponse(status: .fulfilled, result: result) }

    static func denied() -> DeviceRequestResponse {
        DeviceRequestResponse(status: .denied, error: DeviceRequestError(code: .permissionDenied, message: "the person said no"))
    }

    static func failed(_ message: String) -> DeviceRequestResponse {
        DeviceRequestResponse(status: .failed, error: DeviceRequestError(code: .unavailable, message: message))
    }

    /// What this phone tells the hub: its location, off once the person said never.
    static func capability(_ choice: LocationChoice) -> DeviceCapability {
        DeviceCapability(kind: .location, enabled: choice != .never)
    }

    static func request(from envelope: Envelope) -> DeviceRequest? {
        guard envelope.event == "request.created", envelope.namespace == "/rt/devices",
              let object = try? JSONSerialization.jsonObject(with: envelope.payload) as? [String: Any],
              let raw = object["request"], let data = try? JSONSerialization.data(withJSONObject: raw)
        else { return nil }
        return try? HubJSON.decoder.decode(DeviceRequest.self, from: data)
    }

    static let choiceKey = Product.storagePrefix + "location.choice"

    static func storedChoice(_ defaults: UserDefaults = .standard) -> LocationChoice {
        defaults.string(forKey: choiceKey).flatMap(LocationChoice.init(rawValue:)) ?? .ask
    }
}

/// One reading of where the phone is, asking iOS for the permission when it has not yet.
final class PhoneLocator: NSObject, CLLocationManagerDelegate {
    private let manager = CLLocationManager()
    private var waiting: CheckedContinuation<CLLocation?, Never>?
    private var authorizing: CheckedContinuation<Void, Never>?

    override init() {
        super.init()
        manager.delegate = self
        manager.desiredAccuracy = kCLLocationAccuracyHundredMeters
    }

    @MainActor
    func current() async -> CLLocation? {
        if manager.authorizationStatus == .notDetermined {
            await withCheckedContinuation { continuation in
                authorizing = continuation
                manager.requestWhenInUseAuthorization()
            }
        }
        guard manager.authorizationStatus == .authorizedWhenInUse || manager.authorizationStatus == .authorizedAlways else { return nil }
        return await withCheckedContinuation { continuation in
            waiting = continuation
            manager.requestLocation()
        }
    }

    func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        guard manager.authorizationStatus != .notDetermined else { return }
        authorizing?.resume()
        authorizing = nil
    }

    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        waiting?.resume(returning: locations.last)
        waiting = nil
    }

    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        waiting?.resume(returning: nil)
        waiting = nil
    }
}

@MainActor
@Observable
final class LocationRequests {
    static let shared = LocationRequests()

    private(set) var waiting: [DeviceRequest] = []
    private(set) var choice: LocationChoice = Locating.storedChoice()
    @ObservationIgnored private weak var app: AppModel?
    @ObservationIgnored private var namespace: RealtimeNamespace?
    @ObservationIgnored private var listener: UUID?
    @ObservationIgnored private let locator = PhoneLocator()

    func start(app: AppModel) {
        self.app = app
        if namespace == nil {
            // This device's own requests reach this socket only (`device:<id>`).
            let devices = app.realtime.namespace("/rt/devices") { [weak app] in await app?.handshake(all: true) ?? [:] }
            namespace = devices
            listener = devices.onEvent { [weak self] _, argument in
                guard let envelope = Envelope.parse(argument), let request = Locating.request(from: envelope) else { return }
                Task { await self?.heard(request) }
            }
        }
        Task { await catchUp() }
    }

    func set(_ next: LocationChoice) {
        choice = next
        UserDefaults.standard.set(next.rawValue, forKey: Locating.choiceKey)
        Task { await app?.reportLocationChoice() }
    }

    func heard(_ request: DeviceRequest) async {
        switch Locating.next(request, choice: choice) {
        case .ask:
            if !waiting.contains(where: { $0.id == request.id }) { waiting.append(request) }
        case .answer:
            await answer(request)
        case .deny:
            await respond(request, Locating.denied())
        case .ignore:
            break
        }
    }

    /// Requests made while the app was closed.
    func catchUp() async {
        guard let app, let device = await app.api.keeper.credentials?.deviceID else { return }
        let profile = app.currentProfile
        let pending = try? await app.api.call {
            try await DevicesAPI.devicesListRequests(xHubProfile: profile, deviceId: device, status: .pending, limit: 20, apiConfiguration: $0)
        }.items
        for request in pending ?? [] { await heard(request) }
    }

    func decide(_ request: DeviceRequest, allow: Bool, remember: Bool) async {
        waiting.removeAll { $0.id == request.id }
        if remember { set(allow ? .always : .never) }
        if allow { await answer(request) } else { await respond(request, Locating.denied()) }
    }

    private func answer(_ request: DeviceRequest) async {
        guard let place = await locator.current() else {
            await respond(request, Locating.failed("the phone could not read its location"))
            return
        }
        let result = Locating.result(latitude: place.coordinate.latitude, longitude: place.coordinate.longitude, accuracy: place.horizontalAccuracy, at: place.timestamp)
        await respond(request, Locating.fulfilled(result))
    }

    private func respond(_ request: DeviceRequest, _ response: DeviceRequestResponse) async {
        guard let app else { return }
        let profile = request.profile, id = request.id
        _ = try? await app.api.call {
            try await DevicesAPI.devicesRespondRequest(xHubProfile: profile, requestId: id, deviceRequestResponse: response, apiConfiguration: $0)
        }
    }
}

extension AppModel {
    /// Tells the hub whether this phone answers location requests (after the person's choice).
    func reportLocationChoice() async {
        guard let device = await api.keeper.credentials?.deviceID else { return }
        let patch = DevicePatch(capabilities: [Locating.capability(LocationRequests.shared.choice)])
        _ = try? await api.call { try await DevicesAPI.devicesUpdate(deviceId: device, devicePatch: patch, apiConfiguration: $0) }
    }
}

/// The consent sheet: who asks and why, and the three answers.
struct LocationConsentSheet: View {
    let request: DeviceRequest
    @Environment(\.l10n) private var l10n
    @State private var busy = false

    var body: some View {
        VStack(alignment: .leading, spacing: Space.s3) {
            HStack(spacing: Space.s2) {
                LucideIcon(.mapPin, size: 22).foregroundStyle(Tone.accent)
                Text(l10n("locate.title")).font(.system(size: FontSize.sizeLg, weight: .semibold))
            }
            Text(l10n("locate.body")).font(.system(size: FontSize.sizeSm)).foregroundStyle(Tone.textMuted)
            if let why = request.purpose, !why.isEmpty, why != "devices.locate" {
                Text(l10n("locate.why", ["why": why])).font(.system(size: FontSize.sizeSm)).contentDirection(of: why)
            }
            Button {
                decide(allow: true, remember: true)
            } label: {
                LucideLabel(l10n("locate.always"), icon: .mapPin, size: 16).frame(maxWidth: .infinity)
            }
            .buttonStyle(.borderedProminent).tint(Tone.accent)
            .accessibilityIdentifier("locate.always")
            Button { decide(allow: true, remember: false) } label: { Text(l10n("locate.once")).frame(maxWidth: .infinity) }
                .buttonStyle(.bordered)
                .accessibilityIdentifier("locate.once")
            Button(role: .destructive) { decide(allow: false, remember: true) } label: { Text(l10n("locate.never")).frame(maxWidth: .infinity) }
                .accessibilityIdentifier("locate.never")
        }
        .disabled(busy)
        .padding(Space.s4)
        .presentationDetents([.medium])
        .interactiveDismissDisabled(busy)
    }

    private func decide(allow: Bool, remember: Bool) {
        busy = true
        let request = request
        Task { await LocationRequests.shared.decide(request, allow: allow, remember: remember) }
    }
}

/// This device's row: ask, always, or never.
struct LocationChoiceSection: View {
    @Environment(\.l10n) private var l10n

    var body: some View {
        let requests = LocationRequests.shared
        Section {
            Picker(l10n("locate.heading"), selection: Binding(get: { requests.choice }, set: { requests.set($0) })) {
                Text(l10n("locate.ask")).tag(LocationChoice.ask)
                Text(l10n("locate.choice_always")).tag(LocationChoice.always)
                Text(l10n("locate.choice_never")).tag(LocationChoice.never)
            }
            .pickerStyle(.segmented)
            .accessibilityIdentifier("locate.choice")
        } header: {
            Text(l10n("locate.heading"))
        } footer: {
            Text(l10n("locate.hint"))
        }
    }
}

File: apps/android/app/src/main/java/hub/core/android/phone/Locate.kt
package hub.core.android.phone

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationManager
import android.os.Build
import android.os.CancellationSignal
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.setValue
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.content.ContextCompat
import hub.core.android.R
import hub.core.android.data.HubApis
import hub.core.android.data.StoredSession
import hub.core.android.data.hubCall
import hub.core.android.generated.FontTokens
import hub.core.android.graph
import hub.core.android.realtime.DEVICES_NAMESPACE
import hub.core.android.realtime.Envelope
import hub.core.android.ui.kit.ButtonKind
import hub.core.android.ui.kit.ControlSize
import hub.core.android.ui.kit.HubButton
import hub.core.android.ui.kit.HubDialog
import hub.core.android.ui.kit.Lucide
import hub.core.android.ui.theme.LocalTokens
import hub.core.client.infrastructure.Serializer
import hub.core.client.model.CapabilityKind
import hub.core.client.model.DeviceCapability
import hub.core.client.model.DeviceRequest
import hub.core.client.model.DeviceRequestError
import hub.core.client.model.DeviceRequestResponse
import hub.core.client.model.DeviceRequestStatus
import java.time.Instant
import java.time.OffsetDateTime
import java.time.ZoneOffset
import kotlin.coroutines.resume
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withTimeoutOrNull
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive

/** What the person answered once for agents asking where this phone is (contract decision §105). */
enum class LocationChoice { ASK, ALWAYS, NEVER }

/** The rules of answering a location request, apart from Android so they are tested. */
object Locating {
    /** What to do with a request: ask the person, answer it, or say no at once. */
    enum class Next { ASK, ANSWER, DENY, IGNORE }

    fun next(request: DeviceRequest, choice: LocationChoice, now: OffsetDateTime = OffsetDateTime.now()): Next = when {
        request.capability != CapabilityKind.LOCATION || request.status != DeviceRequestStatus.PENDING -> Next.IGNORE
        request.expiresAt.isBefore(now) -> Next.IGNORE
        choice == LocationChoice.NEVER -> Next.DENY
        choice == LocationChoice.ALWAYS -> Next.ANSWER
        else -> Next.ASK
    }

    /** The one shape a location has (§14): latitude, longitude, accuracy in metres, when it was read. */
    fun result(latitude: Double, longitude: Double, accuracyM: Double, capturedAtMs: Long): Map<String, JsonElement> = mapOf(
        "latitude" to JsonPrimitive(latitude),
        "longitude" to JsonPrimitive(longitude),
        "accuracy_m" to JsonPrimitive(accuracyM.coerceAtLeast(0.0)),
        "captured_at" to JsonPrimitive(Instant.ofEpochMilli(capturedAtMs).atOffset(ZoneOffset.UTC).withNano(0).toString()),
    )

    fun fulfilled(result: Map<String, JsonElement>) = DeviceRequestResponse(status = DeviceRequestResponse.Status.FULFILLED, result = result)

    fun denied() = DeviceRequestResponse(
        status = DeviceRequestResponse.Status.DENIED,
        error = DeviceRequestError(code = DeviceRequestError.Code.PERMISSION_DENIED, message = "the person said no"),
    )

    fun failed(message: String) = DeviceRequestResponse(
        status = DeviceRequestResponse.Status.FAILED,
        error = DeviceRequestError(code = DeviceRequestError.Code.UNAVAILABLE, message = message),
    )

    /** What this phone tells the hub it can do: its location, off once the person said never. */
    fun capability(choice: LocationChoice) = DeviceCapability(kind = CapabilityKind.LOCATION, enabled = choice != LocationChoice.NEVER)

    /** A `request.created` envelope's request, when it is one. */
    fun requestOf(e: Envelope): DeviceRequest? {
        if (e.namespace != DEVICES_NAMESPACE || e.event != "request.created") return null
        val raw = e.payload["request"] ?: return null
        return runCatching { Serializer.kotlinxSerializationJson.decodeFromJsonElement(DeviceRequest.serializer(), raw) }.getOrNull()
    }
}

/** The person's answer, kept on the phone (This device changes it). */
class LocationChoices(private val prefs: SharedPreferences) {
    private val _choice = MutableStateFlow(
        runCatching { LocationChoice.valueOf(prefs.getString(KEY, null) ?: "") }.getOrDefault(LocationChoice.ASK),
    )
    val choice: StateFlow<LocationChoice> = _choice.asStateFlow()

    fun set(choice: LocationChoice) {
        prefs.edit().putString(KEY, choice.name).apply()
        _choice.value = choice
    }

    private companion object { const val KEY = "location_choice" }
}

/**
 * The location requests this phone has to answer: heard on `/rt/devices` (`request.created`,
 * sent to this device only) and caught up with `listRequests?status=pending` when the app comes
 * back. Those that need the person wait in [waiting] for the consent dialog.
 */
class LocationRequests(
    private val apis: () -> Pair<HubApis, StoredSession>?,
    private val choices: LocationChoices,
    private val locate: suspend () -> Map<String, JsonElement>?,
) {
    private val _waiting = MutableStateFlow<List<DeviceRequest>>(emptyList())
    val waiting: StateFlow<List<DeviceRequest>> = _waiting.asStateFlow()

    suspend fun heard(request: DeviceRequest) {
        when (Locating.next(request, choices.choice.value)) {
            Locating.Next.ASK -> _waiting.update { list -> if (list.any { it.id == request.id }) list else list + request }
            Locating.Next.ANSWER -> answer(request)
            Locating.Next.DENY -> respond(request, Locating.denied())
            Locating.Next.IGNORE -> Unit
        }
    }

    /** Pending requests for this phone the socket may have missed (the app was closed). */
    suspend fun catchUp() {
        val (api, session) = apis() ?: return
        val device = session.deviceId ?: return
        hubCall { api.devices.devicesListRequests(session.profile, deviceId = device, status = DeviceRequestStatus.PENDING, limit = 20).items }
            .onSuccess { list -> list.forEach { heard(it) } }
    }

    /** The person answered the dialog: yes (once or every time) reads the place; no says so. */
    suspend fun decide(request: DeviceRequest, allow: Boolean, remember: Boolean) {
        _waiting.update { list -> list.filterNot { it.id == request.id } }
        if (remember) choices.set(if (allow) LocationChoice.ALWAYS else LocationChoice.NEVER)
        if (allow) answer(request) else respond(request, Locating.denied())
    }

    private suspend fun answer(request: DeviceRequest) {
        val place = runCatching { locate() }.getOrNull()
        respond(request, place?.let(Locating::fulfilled) ?: Locating.failed("the phone could not read its location"))
    }

    private suspend fun respond(request: DeviceRequest, response: DeviceRequestResponse) {
        val (api, _) = apis() ?: return
        hubCall { api.devices.devicesRespondRequest(request.profile, request.id, response) }
    }
}

/** This phone's current place, from Android's own location service (no Play services needed). */
object PhoneLocation {
    fun allowed(context: Context): Boolean =
        ContextCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED ||
            ContextCompat.checkSelfPermission(context, Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED

    @SuppressLint("MissingPermission") // checked by allowed()
    suspend fun read(context: Context): Map<String, JsonElement>? {
        if (!allowed(context)) return null
        val manager = context.getSystemService(Context.LOCATION_SERVICE) as? LocationManager ?: return null
        val provider = listOf(LocationManager.GPS_PROVIDER, LocationManager.NETWORK_PROVIDER).firstOrNull { runCatching { manager.isProviderEnabled(it) }.getOrDefault(false) }
        val fresh: Location? = if (provider != null && Build.VERSION.SDK_INT >= 30) {
            withTimeoutOrNull(20_000) {
                suspendCancellableCoroutine { cont ->
                    val cancel = CancellationSignal()
                    cont.invokeOnCancellation { cancel.cancel() }
                    manager.getCurrentLocation(provider, cancel, context.mainExecutor) { cont.resume(it) }
                }
            }
        } else null
        val location = fresh ?: listOf(LocationManager.GPS_PROVIDER, LocationManager.NETWORK_PROVIDER, LocationManager.PASSIVE_PROVIDER)
            .mapNotNull { runCatching { manager.getLastKnownLocation(it) }.getOrNull() }
            .maxByOrNull { it.time }
            ?: return null
        return Locating.result(location.latitude, location.longitude, location.accuracy.toDouble(), location.time)
    }
}

/**
 * The consent dialog, the first time an agent asks (and every time while the person keeps
 * «Only this time»): who asks and why, then Android's own location permission when needed.
 */
@Composable
fun LocationConsent() {
    val context = LocalContext.current
    val graph = context.graph
    val scope = rememberCoroutineScope()
    val t = LocalTokens.current
    val waiting by graph.locations.waiting.collectAsState()
    val request = waiting.firstOrNull() ?: return
    var pending by androidx.compose.runtime.remember { androidx.compose.runtime.mutableStateOf<Pair<Boolean, Boolean>?>(null) }
    val permission = rememberLauncherForActivityResult(ActivityResultContracts.RequestMultiplePermissions()) { granted ->
        val (allow, remember) = pending ?: return@rememberLauncherForActivityResult
        scope.launch { graph.locations.decide(request, allow && granted.values.any { it }, remember && granted.values.any { it }) }
    }
    fun go(allow: Boolean, remember: Boolean) {
        if (allow && !PhoneLocation.allowed(context)) {
            pending = allow to remember
            permission.launch(arrayOf(Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION))
        } else {
            scope.launch { graph.locations.decide(request, allow, remember) }
        }
    }
    HubDialog({ go(allow = false, remember = false) }, stringResource(R.string.locate_title)) {
        Column(Modifier.fillMaxWidth().testTag("locate.dialog"), verticalArrangement = Arrangement.spacedBy(8.dp)) {
            Text(stringResource(R.string.locate_body), fontSize = FontTokens.sizeSm.sp, color = t.textMuted)
            request.purpose?.takeIf { it.isNotBlank() && it != "devices.locate" }?.let {
                Text(stringResource(R.string.locate_why, it), fontSize = FontTokens.sizeSm.sp)
            }
            HubButton(stringResource(R.string.locate_always), { go(allow = true, remember = true) }, icon = Lucide.MapPin, fill = true, size = ControlSize.Md, modifier = Modifier.fillMaxWidth().testTag("locate.always"))
            HubButton(stringResource(R.string.locate_once), { go(allow = true, remember = false) }, kind = ButtonKind.Secondary, fill = true, size = ControlSize.Md, modifier = Modifier.fillMaxWidth().testTag("locate.once"))
            HubButton(stringResource(R.string.locate_never), { go(allow = false, remember = true) }, kind = ButtonKind.Ghost, fill = true, size = ControlSize.Md, modifier = Modifier.fillMaxWidth().testTag("locate.never"))
        }
    }
}

